### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by mvl22**
These changes have been focussed on the charts and the layers, moving ever closer to generalised definitions.

Overall, more of the refactoring objectives have been completed; see: #125.

### CODE REVIEW

**Comment by mvl22** [comment]
These changes are focussed on the charts and the layers, moving ever closer to generalised definitions.

Overall, more of the refactoring objectives have been completed; see: #125.

File: css/main.css [hunk 1]
```diff
@@ -357,80 +359,35 @@ p.help-end {
...
-  -webkit-user-select: none;
-  -khtml-user-select: none;
-  -moz-user-select: none;
-  -ms-user-select: none;
-  */
   user-select: none;
 }
 
-
-.basemap label::before {
```
Comment (Robinlovelace) [inline css/main.css:391]: Goodbye hardcoded values, hello generalisation1

File: js/layer_control.js [hunk 2]
```diff
@@ -584,106 +602,18 @@ function getStyleColumn (layerId)
...
-  
-  // Set building colour
-  const buildingColour = getBuildingsColour ();
-  map.setPaintProperty ('dasymetric', 'fill-extrusion-color', (buildingColour || '#9c9898'));
+  // Create the legend HTML
+  // #!# Should be a list, not nested divs
+  let legendHtml = '<div class="l_r">';
+  selected = (legendColours.hasOwnProperty (selected) ? selected : '_');
+  legendColours[selected].forEach(legendColour => {
+    legendHtml += `<div class="lb"><span style="background-color: ${legendColour[1]}"></span>${legendColour[0]}</div>`;
```
Comment (Robinlovelace) [inline js/layer_control.js:612]: So much more concise! Not sure what the `$` syntax does here but I like it.

**Review by Robinlovelace — APPROVED** [review]
This is great work @mvl22 many thanks, sounds like you've boshed the bulk of the refactoring work.

@mem48 do you want to comment on this? Let's get it merged, I'll plan to merge this latest batch of changes by end of day if there are no objections to doing so.

**Comment by mvl22** [comment]
> This is great work @mvl22 many thanks, sounds like you've boshed the bulk of the refactoring work.

I've got through the bulk of the low-level detail stuff, in particular splitting out generalisable stuff from definitions and layer-specific behaviour. Almost everything is now within functions, so scope is easier/safer to handle now. I have still some fiddling around to standardise the HTML attribute namings.

The main change coming next is combining all the logic code into a single unified class, with a settings file and definitions files provided to that, and re-plumbing non-generic layer-specific handling as callbacks. Then the refactoring is basically done.

Then I'll put in a few styling changes.

**Pull request merged**